docs/source/python/api/dataset.rst

```diff
@@ -0,0 +1,53 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. currentmodule:: pyarrow.dataset
+
+.. _api.dataset:
+
+Datasets
+========
+
+Factory functions
+-----------------
+
+.. autosummary::
+   :toctree: ../generated/
+
+   source
+   dataset
+   partitioning
+
+Classes
+-------
+
+.. autosummary::
+   :toctree: ../generated/
+
+   FileFormat
+   ParquetFileFormat
+   Partitioning
+   PartitioningFactory
+   DirectoryPartitioning
+   HivePartitioning
+   Source
+   FileSystemSource
+   FileSystemFactoryOptions
+   FileSystemSourceFactory
+   Dataset
+   ScannerBuilder
+   Scanner
```

Comment:
I am not fully sure which classes we should include here and which not, since there is still an option about removing some of those. 

(initially for 0.16 we could also start with only documenting the high-level functions)

Reply:
@jorisvandenbossche agree, perhaps we shouldn't even expose the classes under `pyarrow.dataset`. If someone wants to use them they are still available under `pyarrow._dataset` which warns the user about it experimental nature.

Reply:
Yes, that's also certainly an option